dags/xcom_gcs_ds_k8sExecutor.py

```diff
@@ -0,0 +1,220 @@
+
+        bq = BigQueryHook()
+        sql = """
+        SELECT * FROM `bigquery-public-data.ml_datasets.census_adult_income`
+        """
+
+        return bq.get_pandas_df(sql=sql, dialect='standard')
+
+
+
+    @task
+    def preprocessing(df: pd.DataFrame):
+        """Clean Data and prepare for feature engineering
+        
+        Returns pandas dataframe via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- Raw data pulled from BigQuery to be processed. 
+        """
+
+        df.dropna(inplace=True)
+        df.drop_duplicates(inplace=True)
+
+        # Clean Categorical Variables (strings)
+        cols = df.columns
+        for col in cols:
+            if df.dtypes[col]=='object':
+                df[col] =df[col].apply(lambda x: x.rstrip().lstrip())
+
+
+        # Rename up '?' values as 'Unknown'
+        df['workclass'] = df['workclass'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['occupation'] = df['occupation'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['native_country'] = df['native_country'].apply(lambda x: 'Unknown' if x == '?' else x)
+
+
+        # Drop Extra/Unused Columns
+        df.drop(columns=['education_num', 'relationship', 'functional_weight'], inplace=True)
+
+        return df
+
+    @task
+    def feature_engineering(df: pd.DataFrame):
+        """Feature engineering step
+        
+        Returns pandas dataframe via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- data from previous step pulled from BigQuery to be processed. 
+        """
+
+        
+        # Onehot encoding 
+        df = pd.get_dummies(df, prefix='workclass', columns=['workclass'])
+        df = pd.get_dummies(df, prefix='education', columns=['education'])
+        df = pd.get_dummies(df, prefix='occupation', columns=['occupation'])
+        df = pd.get_dummies(df, prefix='race', columns=['race'])
+        df = pd.get_dummies(df, prefix='sex', columns=['sex'])
+        df = pd.get_dummies(df, prefix='income_bracket', columns=['income_bracket'])
+        df = pd.get_dummies(df, prefix='native_country', columns=['native_country'])
+
+
+        # Bin Ages
+        df['age_bins'] = pd.cut(x=df['age'], bins=[16,29,39,49,59,100], labels=[1, 2, 3, 4, 5])
+
+
+        # Dependent Variable
+        df['never_married'] = df['marital_status'].apply(lambda x: 1 if x == 'Never-married' else 0) 
+
+
+        # Drop redundant colulmn
+        df.drop(columns=['income_bracket_<=50K', 'marital_status', 'age'], inplace=True)
+
+        return df
+
+
+    @task(executor_config=modeling_config)
+    def train(df: pd.DataFrame):
+        """Train and validate model
+        
+        Returns accuracy score via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- data from previous step pulled from BigQuery to be processed. 
+        """
+
+        y = df['never_married'].values
+        X = df.drop(columns=['never_married']).values
```

Comment:
I'd pull this step into `feature_engineering` as well, so that what that task returns is X and y -- then `train` and `fit` could be totally agnostic to actual attributes of the data, namely the 'never_married' column name.

Reply:
@santonaastronomer To output two XComs from feature_engineering and then read them  back in I would have to output them as a JSON. There may be a more simpler way that I am overlooking, but something like this...

XCom Output from feature_engineering
```
return {'X': df.drop(columns=['never_married']).to_json(orient='index'), 'y': df['never_married'].to_json(orient='index')}

```

Then read then back in with `io.StringIO` on the downstream tasks. Which would look like this

```
y = pd.read_json(StringIO(df['y']), orient='index').values
X = pd.read_json(StringIO(df['X']), orient='index').values
```

This is not necessarily bad, but wanted to make sure here before committing anything.